### PR TITLE
[openbeats-1] BEATs encoder and tokenizer modules

### PR DESCRIPTION
**Pull request by jaeyeonkim99**
## What did you change?

Ported the BEATs encoder and tokenizer modules from the closed PR #6052.

- Added iterative-pretraining support (masked sequence prediction, EMA teacher path) to `BeatsEncoder`.
- Added BEATs tokenizer modules: `BeatsTokenizer` (trained), `BeatsRandomTokenizer` (random projection), and an `AudioTokenizer` wrapper.
- Added unit tests for the encoder, tokenizers, and shared utils.

This PR contains only the BEATs model components. Task class, training CLI, and the iterative pretraining stage runner are in the follow-up `[openbeats-2]` (#6437). Per-corpus recipes will follow in separate PRs after that.

## Why did you make this change?

BEATs is a widely-used audio SSL backbone. The original PR #6052 became too large to review and merge. Re-porting it as smaller PRs to keep each change reviewable. This one is for the encoder and tokenizer modules, and the follow-up `[openbeats-2]`(#6437) covers the training stack (task class, CLIs, stage runner). Per-corpus recipes will follow in separate PRs.

---

## Is your PR small enough?

No . 10 files, ~1800 lines. The encoder and tokenizer modules are inter-dependent and need to be merged together.

---

## Additional Context

* #6052 

This PR is the original PR for OpenBEATs implementation.

### CODE REVIEW

**Comment by sw005320** [comment]
@popcornell, can you review this PR?

File: espnet2/beats/encoder.py [hunk 1]
```diff
@@ -141,10 +162,11 @@ def __init__(
         specaug_config: Optional[Dict] = None,
         add_positional_information: bool = False,
         max_positions: Optional[int] = None,
-        fbank_mean: float = 15.41663,
-        fbank_std: float = 6.55582,
+        fbank_mean: float = 15.29130,
+        fbank_std: float = 5.90532,
```
Comment (Shikhar-S) [inline espnet2/beats/encoder.py]: nit: Check appendix A https://arxiv.org/pdf/2212.09058
Original values were based off of the BEATs checkpoint, new ones are over our copy of Audioset on delta, used in OpenBeats.
Maybe good to just add a comment specifying the above.

File: espnet2/asr/encoder/beats_encoder.py [hunk 2]
```diff
@@ -341,79 +395,190 @@ def forward(
...
+            not skip_fbank_extraction
+            and self.min_input_length_at_16khz
+            and source.size(1) < self.min_input_length_at_16khz
+        ):
+            # Only executed for raw waveform input
             logging.warning(
                 f"Input shape: {source.shape}. This is less than"
                 f" the minimum size of {self.min_input_length_at_16khz}."
             )
             # repeat the input to make it at least min_length
```
Comment (Shikhar-S) [inline espnet2/asr/encoder/beats_encoder.py]: Another option is to set an assert here, the caller should ensure min length constraints. This means the downstream recipe code would have to handle such manipulation.

File: espnet2/speechlm/tokenizer/random_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from torch import nn
+from torch.linalg import vector_norm
+
+
+class RandomProjectionQuantizer(nn.Module):
+    def __init__(self, dim, codebook_size, codebook_dim, norm=True, seed=45):
+        super().__init__()
+        # Random projection layer
+        self.random_projection = nn.Linear(dim, codebook_dim, bias=False)
+        # nn.init.xavier_uniform_(self.random_projection.weight)
```
Comment (Shikhar-S) [inline espnet2/speechlm/tokenizer/random_tokenizer.py]: nit: delete?

**Review by Shikhar-S — COMMENTED** [review]
Thanks @jaeyeonkim99 for working on this. Added some minor comments, overall lgtm!

File: espnet2/beats/encoder.py (hunk 1 above)
Comment (jaeyeonkim99) [reply espnet2/beats/encoder.py]: Added a comment pointing to Appendix A for the original BEATs stats. Fixed in [72da469](https://github.com/espnet/espnet/pull/6436/commits/72da469f7cbf0d6e2eed29d78d418ca96ad4ba59).

File: espnet2/speechlm/tokenizer/random_tokenizer.py (hunk 3 above)
Comment (jaeyeonkim99) [reply espnet2/speechlm/tokenizer/random_tokenizer.py]: Deleted in [17e7633](https://github.com/espnet/espnet/pull/6436/commits/17e7633800fae4b67487dcf472fae4e5e70cf21d).

File: espnet2/asr/encoder/beats_encoder.py [hunk 4]
```diff
@@ -341,79 +395,190 @@ def forward(
...
-            source = torch.cat(
-                [source] * (self.min_input_length_at_16khz // source.size(1) + 1), dim=1
-            )
+            repeat_factor = self.min_input_length_at_16khz // source.size(1) + 1
+            source = torch.cat([source] * repeat_factor, dim=1)
+            padding_mask = torch.cat([padding_mask] * repeat_factor, dim=1)
```
[the remark this answers is not in the document]
Comment (jaeyeonkim99) [reply espnet2/asr/encoder/beats_encoder.py]: Applied: Change to pad the `source` with mask and make corresponding `padding_mask` to be True. Fixed in [210ad98](https://github.com/espnet/espnet/pull/6436/commits/210ad98801216e0cd9a2010771b3efa258149a70).

**Comment by sw005320** [comment]
Can you fix this CI issue?
https://github.com/espnet/espnet/actions/runs/27006024589/job/79697666834?pr=6436

**Comment by jaeyeonkim99** [comment]
> Can you fix this CI issue? https://github.com/espnet/espnet/actions/runs/27006024589/job/79697666834?pr=6436

Just fixed it!

File: espnet2/asr/encoder/beats_encoder.py [hunk 5]
```diff
@@ -38,6 +39,14 @@
...
 from espnet2.asr.specaug.specaug import SpecAug
 from espnet2.legacy.nets.pytorch_backend.nets_utils import make_pad_mask, roll_tensor
+from espnet2.speechlm.tokenizer.beats_utils import (
+    forward_padding_mask_conv,
+    freeze_conv_module,
+)
+
+is_torch_v25_to_v26 = V(torch.__version__) >= V("2.5.0") and V(torch.__version__) <= V(
+    "2.6.0"
+)
```
Comment (sw005320) [inline espnet2/asr/encoder/beats_encoder.py]: Now, espnet does not support 2.6.0 (minimum 2.9.1, see https://github.com/espnet/espnet#espnet-end-to-end-speech-processing-toolkit).
It's up to you, but we may drop this.

Comment (sw005320) [reply espnet2/asr/encoder/beats_encoder.py]: Removing this condition simplifies the following code.

File: espnet2/speechlm/tokenizer/abs_tokenizer.py
Comment (sw005320) [inline espnet2/speechlm/tokenizer/abs_tokenizer.py:1]: As I said, we can move this under `espnet2/beats`
I was checking the dependency and there is not so much dependencies from `espnet2/speechlm`
This will improve the modularity of your implementation.

File: espnet2/speechlm/tokenizer/beats_tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+from packaging.version import parse as V
+from torch.nn import LayerNorm
+
+if V(torch.__version__) >= V("1.6.0"):
+    from torch.cuda.amp import autocast
+else:
+    # Nothing to do if torch<1.6.0
+    @contextmanager
+    def autocast(enabled=True):
+        yield
```
Comment (sw005320) [inline espnet2/speechlm/tokenizer/beats_tokenizer.py]: We don't need this anymore.

File: espnet2/beats/tokenizer.py [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+
+class BeatsTokenizer(BeatsEncoder):
+    def __init__(
+        self,
+        beats_tokenizer_ckpt_path: str = None,
+        tokenizer_config: Optional[Dict] = None,
+        max_layer: int = None,
+        use_weighted_representation: bool = False,
+        fbank_mean: float = 15.29130,
+        fbank_std: float = 5.90532,
```
Comment (sw005320) [inline espnet2/beats/tokenizer.py]: I'm concerned that it has such default values and is provided in multiple places (e.g., lines 165/166 in espnet2/asr/encoder/beats_encoder.py)
Are they actually used?
If so, there might be some inconsistency if we forget to change both.

Is there a way to avoid providing a detailed value for different places?

File: espnet2/beats/encoder.py
Comment (sw005320) [inline espnet2/beats/encoder.py:1]: ditto.
Maybe we can move this from `espnet2/asr/encoder/beats_encoder.py` to `espnet2/beats/encoder/encoder.py`?
This will be better in terms of modularity.

**Review by sw005320 — COMMENTED** [review]
This looks good to me overall.
Can you try to move the`tokenizer` module from `speechlm` to `beats`?
It seems that it is doable.
With this, the modularity would be improved!

File: espnet2/beats/audio_tokenizer.py
Comment (sw005320) [inline espnet2/beats/audio_tokenizer.py:1]: May you also move this under `beats/`?
If there is a particular reason, you can keep them in this location, but I'd recommend you put the comments (in the header of this file of why we need to lcate this even it is used in the beats model

File: espnet2/tasks/asr.py [hunk 8]
```diff
@@ -74,6 +73,7 @@
 from espnet2.asr.specaug.abs_specaug import AbsSpecAug
 from espnet2.asr.specaug.specaug import SpecAug
 from espnet2.asr_transducer.joint_network import JointNetwork
+from espnet2.beats.encoder import BeatsEncoder
```
Comment (sw005320) [inline espnet2/tasks/asr.py:76]: Do you need this?
I thought that we don't have to go through the ASR task

File: espnet3/systems/asr/task.py [hunk 9]
```diff
@@ -79,6 +78,7 @@
 from espnet2.asr.specaug.abs_specaug import AbsSpecAug
 from espnet2.asr.specaug.specaug import SpecAug
 from espnet2.asr_transducer.joint_network import JointNetwork
+from espnet2.beats.encoder import BeatsEncoder
```
Comment (sw005320) [inline espnet3/systems/asr/task.py]: ditto

**Comment by sw005320** [comment]
I think once the current refactoring is finished, we can merge this PR.
We can move to #6437

**Comment by sw005320** [comment]
Can you check the CI error?
This seems to come from your change.

**Comment by sw005320** [comment]
Does this come from the timeout issue?
If so, I recommend you reduce the network architecture very very small.
The CI is based on (slow) CPUs.
You can check the other test.

**Comment by jaeyeonkim99** [comment]
> Does this come from the timeout issue?
> If so, I recommend you reduce the network architecture very very small.
> The CI is based on (slow) CPUs.
> You can check the other test.

Thanks for suggestion. The cause was memory. CI failed at `alloc_cpu` for exactly 25,600,320,001 bytes = (160001)² bool, which is the `torch.ones(maxlen+1, maxlen+1)` helper `espnet2/legacy/nets_utils.py::_make_pad_mask_traceable` (raw 10 s waveform, maxlen=160k → ~25.6 GB). Tests already use a small config (`encoder_layers=2`). 

I fixed this issue in [89f0ee1](https://github.com/espnet/espnet/pull/6436/commits/89f0ee1c19): routed BEATs' five pad-mask sites through a linear-memory helper, bit-identical output, but using memory of 25.6 GB → ~320 KB. The previously failed`test_python_espnet2 (3.12, *)` is now green.

The `execution_timeout(60)` markers ([d4c54fe](https://github.com/espnet/espnet/pull/6436/commits/d4c54fe13b), [0607247](https://github.com/espnet/espnet/pull/6436/commits/06072477b6)) are separate, as backward-pass tests still need >10 s on CI CPU.

**Comment by sw005320** [comment]
Thanks!

**Comment by sw005320** [comment]
BTW, since it is a test, you might use a shorter sound (e.g., 1 second could be enough?).
The test does not need to be the real setup

**Comment by jaeyeonkim99** [comment]
> BTW, since it is a test, you might use a shorter sound (e.g., 1 second could be enough?). The test does not need to be the real setup

Thank you for the suggestion. The 10 s clip was just inherited from the original OpenBEATs implementation, and the other BEATs tokenizer tests already use 1 s inputs. Following your suggestion, I shortened the two random-tokenizer tests to ~1 s in [e707218](https://github.com/espnet/espnet/pull/6436/commits/e7072182a9).

**Comment by sw005320** [comment]
Thanks!
Great job!
Let's go to the next one!
